[TensorExpr][CPU] Fix bool -> int casting

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47374 [NNC] Enable unary op cpu testing
* #47373 [NNC] refactor cuda half support to more general file
* #47372 [NNC] More cpu tests
* #47371 [NNC] Add more CPU Tests
* #47370 [NNC] Fix llvm min lowering for int inputs
* #47369 Add more CPU tests
* **#46951 [TensorExpr][CPU] Fix bool -> int casting**
* #46950 Small changes/cleanup
* #46949 Force LLVM Compilation for CPU Tests

If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set.

Differential Revision: [D24805253](https://our.internmc.facebook.com/intern/diff/D24805253)

Commit:
Fix bool -> int casting

[ghstack-poisoned]

Commit:
Update on "Fix bool -> int casting"




[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
![image](https://user-images.githubusercontent.com/4588111/97375014-f32dba80-1876-11eb-8004-52c863f08ee5.png)

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -826,7 +827,21 @@ void LLVMCodeGenImpl::visit(const Cast* v) {
       }
+
     } else if (dstType->isIntOrIntVectorTy()) {
+      // Bool is represented as 8bits, we need to make sure
```
I find this comment (and the description) a bit misleading. Any of the 8 bits is good-enough as far as true / false goes. We only hit problems when we observe the exact underlying value through a cast to integer.

I'd prefer a phrasing which focuses on the fact that we want true to be exactly 1 (as opposed to any non-zero value), that'd reflect better what's being fixed here.

eellison replied:
How is 
> we're appropriately setting the one bit that is used

different from 

> we want true to be exactly 1 

? I'm also not really sure that `Any of the 8 bits is good-enough as far as true/false goes` is true. Pytorch only uses the one bit. That is why the test was failing previously, because only the second bit was being set for `True + True`.

asuhan replied:
When we add two `True` bools, we still get the correct *boolean* value, `True` on master. I printed the values in the test. It fails because the subsequent integer conversion yielded 2 instead of 1. Reading this comment, my guess would be that we lost all set bits and ended up with a `False` instead of a `True`.

ZolotukhinM replied:
For me the original comment was also clear with regard that "the one bit that is used" meant that just promoting to ints and doing arithmetic there was not enough and led to incorrect results. It might be worth adding the 1+1 example just to make it crystal clear though.

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -826,7 +827,21 @@ void LLVMCodeGenImpl::visit(const Cast* v) {
+      // we're appropriately setting the one bit that is used
+      if (v->dtype().scalar_type() == ScalarType::Bool) {
+        llvm::Constant* constant_zero = llvm::ConstantInt::get(srcType, 0);
+        if (v->src_value()->dtype().lanes() != 1) {
```
Do we have test coverage for the multi-lane case? If not, we should add it.

asuhan replied:
The special case for boolean should be moved to a helper function. We have too many levels of nesting here.

eellison replied:
yes we do for multi-lane case. agreed i can reduce nesting.

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -826,7 +827,21 @@ void LLVMCodeGenImpl::visit(const Cast* v) {
+      if (v->dtype().scalar_type() == ScalarType::Bool) {
+        llvm::Constant* constant_zero = llvm::ConstantInt::get(srcType, 0);
+        if (v->src_value()->dtype().lanes() != 1) {
+          std::vector<llvm::Constant*> const_vector;
```
I think `zeroes` instead of `const_vector` and `zero` instead of `constant_zero` would be both shorter and more descriptive. Knowing it's a zero already implies it's a constant.

asuhan reviewed (changes requested):
LGTM mostly, see inline comments.

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -124,6 +125,7 @@ class LLVMCodeGenImpl : public IRVisitor {
   llvm::Type* dtypeToLLVMPtr(Dtype dtype);
   void emitWrapper(const std::vector<llvm::Type*>& params);
   void emitKernel(Stmt* stmt, const std::vector<llvm::Type*>& params);
+  llvm::Value * toVec(llvm::Value * v, int lanes);
```
Indentation doesn't look right here.

asuhan reviewed (approved):
Looks good assuming indentation is fixed.

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set. 

[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set.

Differential Revision: [D24805253](https://our.internmc.facebook.com/intern/diff/D24805253)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr][CPU] Fix bool -> int casting"


If e.g. we're casting from torch.int -> torch.bool,  previously we would just truncate from int32 -> i8. Since torch.bool has 8 bits but only uses one of them, we need to makes sure that one bit is set.

Differential Revision: [D24805253](https://our.internmc.facebook.com/intern/diff/D24805253)

[ghstack-poisoned]
